src/http.rs

```diff
@@ -0,0 +1,118 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use std::collections::HashMap;
+use std::fmt;
+use std::io::{Read, Write};
+use std::net::TcpStream;
+use std::result::Result;
+
+use native_tls::TlsConnector;
+use url::Url;
+
+#[derive(Debug)]
+pub struct Response {
+    status: u16,
+    headers: HashMap<String, String>,
+    body: Vec<u8>,
+}
+
+impl Response {
+    pub fn header(&self, find: &str) -> Option<&str> {
+        let mut find = find.to_string();
+        find.make_ascii_lowercase();
+        self.headers.get(&find).map(|s| s.as_str())
+    }
+
+    pub fn into_reader(self) -> impl Read + Send {
+        std::io::Cursor::new(self.body)
+    }
+}
+
+#[derive(Debug)]
+pub enum Error {
+    Io(std::io::Error),
+    MalformedResponse,
+    NoHost,
+    NoPort,
+    NoStatus,
+    Status(u16, Response),
+    Tls(Box<dyn std::error::Error + Send + Sync>),
+    UrlParse(url::ParseError),
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+impl std::error::Error for Error {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        None
+    }
+}
+
+pub fn get(url: &str) -> Result<Response, Error> {
+    let url = Url::parse(url).map_err(Error::UrlParse)?;
+    let host = url.host_str().ok_or(Error::NoHost)?;
+    let port = url.port_or_known_default().ok_or(Error::NoPort)?;
```

Comment:
If you pass in a `Url` rather than a `str` then you don't need a bunch of these `Error` variants.

Reply:
I think it just lets us drop one of them, since these methods called on the parsed Url just return `Option`. I've gone with this suggestion for the next revision.